internal/k8s/reconciler/gateway.go

```diff
@@ -464,8 +464,19 @@ func (g *K8sGateway) ensureDeploymentExists(ctx context.Context) error {
 
 	deployment := g.deploymentBuilder.Build()
 	mutated := deployment.DeepCopy()
+
+	currentDeployment, err := g.client.GetDeployment(ctx, types.NamespacedName{Namespace: deployment.Namespace, Name: deployment.Name})
+	if err != nil {
+		return err
+	}
+
+	var currentReplicas *int32
+	if currentDeployment != nil {
+		currentReplicas = currentDeployment.Spec.Replicas
+	}
+	mutated = apigwv1alpha1.MergeDeployment(deployment, mutated, currentReplicas)
```

Comment:
This is existing code, but it doesn't make sense to me why we "merge" a freshly-generated deployment onto a `.DeepCopy()` of itself.

This drew my attention because we're passing information from a third `Deployment` we have in memory as `currentReplicas`, which strikes me as odd.

Based on the context, I would expect `MergeDeployment(deployment, currentDeployment)` so that we're merging a newly-constructed deployment onto the existing one.

Can you speak to this @andrewstucki ?

Reply:
For what its worth, in https://github.com/hashicorp/consul-api-gateway/pull/198 I pulled the logic out of mergedeployment and put it into the build function. I still don't know why this is needed though.

Reply:
So, this could probably be refactored in general, but the idea is that the merge happens on an update, identically to the controller runtime helper that does the same sort of thing: https://github.com/hashicorp/consul-api-gateway/blob/9d0b1d8de9db535edf0c6e8ad33feddba4feb0b4/internal/k8s/gatewayclient/gatewayclient.go#L368

if you dig into the actual implementation of that helper, basically when it detects a change in what exists in k8s v. what exists on disk, it overrides the passed in object (`mutated` in this case), then calls the mutation functions before calling `client.Update`. The `DeepCopy` is done so that on:

1. `Create`, the object gets created as it should, and 
2. On `Update`, when `MergeDeployment` is called, the value is mutated to be reflective of the constructed state.

What this means is that @sarahalsmiller you'll need to put the `MergeDeployment` call back inside of the callback -- that's where any mutations should happen for an `Update` call.